perun-core/src/main/java/cz/metacentrum/perun/core/impl/modules/attributes/urn_perun_user_attribute_def_virt_groupNames.java

```diff
@@ -37,25 +40,28 @@
 	private static final String FRIENDLY_NAME = "groupNames";
 	private static final String A_U_V_GROUP_NAMES = AttributesManager.NS_USER_ATTR_VIRT + ":" + FRIENDLY_NAME;
 
-
 	@Override
 	public Attribute getAttributeValue(PerunSessionImpl sess, User user, AttributeDefinition attributeDefinition) throws InternalErrorException {
 		Attribute attribute = new Attribute(attributeDefinition);
 		List<String> groupNames = new ArrayList<>();
-
-		List<Member> members = sess.getPerunBl().getMembersManagerBl().getMembersByUser(sess, user);
-		Set<String> voNames = new HashSet<>();
-		for (Member member : members) {
-			Vo vo = sess.getPerunBl().getMembersManagerBl().getMemberVo(sess, member);
-			voNames.add(vo.getShortName());
-
-			List<Group> groups = sess.getPerunBl().getGroupsManagerBl().getGroupsWhereMemberIsActive(sess, member);
-			for (Group group : groups) {
-				groupNames.add(vo.getShortName() +":"+ group.getName());
+		List<Map<String, Object>> names = sess.getPerunBl().getDatabaseManagerBl().getJdbcPerunTemplate().queryForList(
```

Comment:
Sorry I didn't realize before. You can use our api object ``Pair<A,B>`` with both types specified as ``String`` and perform standard ``query()`` with passed ``RowMapper``. See ``RegistrarManagerImpl`` class in ``perun-registrar-lib`` module, around line 1302.

And we should wrap the whole query part in try/catch to handle ``EmptyResultDataAccessException`` - means no row selected, we can return empty list as attribute value in this case. Other exceptions should be catched, logged as error and wrapped in ``InternalErrorException``

Thank you.

Reply:
I've done what you requested and used `Pair<>` and ` query()` but because of that change, the test started to fail. It seems like it's ignoring the result of `when(...).thenReturn(...)` even though the call inside ` when()` is the same as the one that is used. I can amend the code and then fix it if you know what the issue is.

Reply:
I'll look into it and let you know. If it won't work, we can use your original solution, since it basically works.

Reply:
It works for me in the test. Catch is, that you need to have "same" parameters when you mock the call as in module. In this case you need static instance of RowMapper in the module and use is also in the test.

In module:
```java
protected static final RowMapper<Pair<String, String>> ROW_MAPPER = new RowMapper<Pair<String, String>>() {
		@Override
		public Pair<String, String> mapRow(ResultSet rs, int rowNum) throws SQLException {
			String voShortName = rs.getString("vo_short_name");
			String groupName = rs.getString("group_name");
			return new Pair<>(voShortName, groupName);
		}
	};
```

In test:
```java
		when(session.getPerunBl().getDatabaseManagerBl().getJdbcPerunTemplate().query("SELECT" +
			" DISTINCT vos.short_name AS vo_short_name, groups.name AS group_name" +
			" FROM" +
			" members" +
			" JOIN vos ON vos.id = members.vo_id AND members.user_id = ?" +
			" JOIN groups_members ON groups_members.member_id = members.id AND groups_members.source_group_status = ?" +
			" JOIN groups ON groups_members.group_id = groups.id" +
			" ORDER BY vo_short_name, group_name", urn_perun_user_attribute_def_virt_groupNames.ROW_MAPPER,
				user.getId(), MemberGroupStatus.VALID.getCode())).thenReturn(dbResult);
```

Reply:
Thanks, done.